### ISSUE TITLE
CopyFilesOverSSH fails on MacOS agent with "Unable to create target folder"

### ISSUE THREAD

**Issue by rogfut** [microsoft/azure-pipelines-tasks#16284]
## Note  
Issues in this repo are for tracking bugs, feature requests and questions for the tasks in this repo 

For a list:  
https://github.com/Microsoft/azure-pipelines-tasks/tree/master/Tasks

If you have an issue or request for the Azure Pipelines service, use developer community instead:  

https://developercommunity.visualstudio.com/spaces/21/index.html )

## Required Information

Entering this information will route you directly to the right team and expedite traction.

**Question, Bug, or Feature?**  
*Type*: Bug

**Enter Task Name**: CopyFilesOverSSH

list here (V# not needed):  
https://github.com/Microsoft/azure-pipelines-tasks/tree/master/Tasks

## Environment
- Server - Azure Pipelines or TFS on-premises? TFS on-premises
    
    - If using TFS on-premises, provide the version: 18.181.31626.1
    
    - If using Azure Pipelines, provide the account name, team project name, build definition name/build number: 

- Agent - Hosted or Private: 
    
    - If using Hosted agent, provide agent queue name:

    - If using private agent, provide the OS of the machine running the agent and the agent version:  Mac OS 11.6.1, Mac OS 11.6, AGENT_VERSION=2.181.2

## Issue Description

When running CopyOverSSH task version 0.180.0 on my private macos vsts agent, the task fails with "Unable to create target folder".  The issue started when we upgraded to TFS on-premises version 18.181.31626.1 with no changes to the pipeline.  The VSTS agent was upgraded from version 2.153.1 to 2.181.2, but if I recall correctly the issue happened on both agent versions.

The last known working version of this task was 0.137.2 on TFS on-prem version 2019.x.x

NOTE: There seems to have been a  bug fixed for this task in [#13507](https://github.com/microsoft/azure-pipelines-tasks/pull/13507) with similar symptoms.  My remote target is a windows machine

### Task logs

2022-05-04T22:30:31.2903750Z ##[debug]Evaluating condition for step: 'Securely copy files to the remote machine'
2022-05-04T22:30:31.2905090Z ##[debug]Evaluating: succeeded()
2022-05-04T22:30:31.2905530Z ##[debug]Evaluating succeeded:
2022-05-04T22:30:31.2906240Z ##[debug]=> True
2022-05-04T22:30:31.2906840Z ##[debug]Result: True
2022-05-04T22:30:31.2907470Z ##[section]Starting: Securely copy files to the remote machine
2022-05-04T22:30:31.2915610Z ==============================================================================
2022-05-04T22:30:31.2915890Z Task         : Copy files over SSH
2022-05-04T22:30:31.2916160Z Description  : Copy files or build artifacts to a remote machine over SSH
2022-05-04T22:30:31.2916410Z Version      : 0.180.0
2022-05-04T22:30:31.2916620Z Author       : Microsoft Corporation
2022-05-04T22:30:31.2916910Z Help         : https://docs.microsoft.com/azure/devops/pipelines/tasks/deploy/copy-files-over-ssh
2022-05-04T22:30:31.2917240Z ==============================================================================
2022-05-04T22:30:31.2967500Z ##[debug]Using node path: /Users/roger/mbp-3/externals/node10/bin/node
2022-05-04T22:30:31.4720730Z ##[debug]agent.TempDirectory=/Users/roger/mbp-3/_w/_temp
2022-05-04T22:30:31.4746500Z ##[debug]loading inputs and endpoints
2022-05-04T22:30:31.4755790Z ##[debug]loading INPUT_SSHENDPOINT
2022-05-04T22:30:31.4776570Z ##[debug]loading INPUT_SOURCEFOLDER
2022-05-04T22:30:31.4779480Z ##[debug]loading INPUT_CONTENTS
2022-05-04T22:30:31.4780620Z ##[debug]loading INPUT_TARGETFOLDER
2022-05-04T22:30:31.4782060Z ##[debug]loading INPUT_ISWINDOWSONTARGET
2022-05-04T22:30:31.4783540Z ##[debug]loading INPUT_CLEANTARGETFOLDER
2022-05-04T22:30:31.4785080Z ##[debug]loading INPUT_READYTIMEOUT
2022-05-04T22:30:31.4786200Z ##[debug]loading INPUT_OVERWRITE
2022-05-04T22:30:31.4787380Z ##[debug]loading INPUT_FAILONEMPTYSOURCE
2022-05-04T22:30:31.4813390Z ##[debug]loading INPUT_FLATTENFOLDERS
2022-05-04T22:30:31.4814720Z ##[debug]loading ENDPOINT_AUTH_95176922-d14c-4134-af49-b195beea74ac
2022-05-04T22:30:31.4815670Z ##[debug]loading ENDPOINT_AUTH_SCHEME_95176922-d14c-4134-af49-b195beea74ac
2022-05-04T22:30:31.4816520Z ##[debug]loading ENDPOINT_AUTH_PARAMETER_95176922-d14c-4134-af49-b195beea74ac_USERNAME
2022-05-04T22:30:31.4817380Z ##[debug]loading ENDPOINT_AUTH_PARAMETER_95176922-d14c-4134-af49-b195beea74ac_PASSWORD
2022-05-04T22:30:31.4818010Z ##[debug]loading ENDPOINT_AUTH_SYSTEMVSSCONNECTION
2022-05-04T22:30:31.4818590Z ##[debug]loading ENDPOINT_AUTH_SCHEME_SYSTEMVSSCONNECTION
2022-05-04T22:30:31.4819180Z ##[debug]loading ENDPOINT_AUTH_PARAMETER_SYSTEMVSSCONNECTION_ACCESSTOKEN
2022-05-04T22:30:31.4819790Z ##[debug]loading SECRET_PROVISIONINGCERTIFICATEPASSWORD
2022-05-04T22:30:31.4820290Z ##[debug]loaded 18
2022-05-04T22:30:31.4820790Z ##[debug]Agent.ProxyUrl=undefined
2022-05-04T22:30:31.4821300Z ##[debug]Agent.CAInfo=undefined
2022-05-04T22:30:31.4821790Z ##[debug]Agent.ClientCert=undefined
2022-05-04T22:30:31.4822300Z ##[debug]Agent.SkipCertValidation=undefined
2022-05-04T22:30:31.6090650Z ##[debug]check path : /Users/roger/mbp-3/_w/_tasks/CopyFilesOverSSH_67cec91b-0351-4c2f-8465-d74b3d2a2d96/0.180.0/task.json
2022-05-04T22:30:31.6093170Z ##[debug]adding resource file: /Users/roger/mbp-3/_w/_tasks/CopyFilesOverSSH_67cec91b-0351-4c2f-8465-d74b3d2a2d96/0.180.0/task.json
2022-05-04T22:30:31.6093980Z ##[debug]system.culture=en-US
2022-05-04T22:30:31.6125490Z ##[debug]sshEndpoint=95176922-d14c-4134-af49-b195beea74ac
2022-05-04T22:30:31.6126840Z ##[debug]95176922-d14c-4134-af49-b195beea74ac auth param username = ***
2022-05-04T22:30:31.6128060Z ##[debug]95176922-d14c-4134-af49-b195beea74ac auth param password = ***
2022-05-04T22:30:31.6128800Z ##[debug]95176922-d14c-4134-af49-b195beea74ac data host = remoteserver
2022-05-04T22:30:31.6129500Z ##[debug]95176922-d14c-4134-af49-b195beea74ac data port = 22
2022-05-04T22:30:31.6130350Z ##[debug]readyTimeout=20000
2022-05-04T22:30:31.6130840Z ##[debug]Using username and password for ssh connection.
2022-05-04T22:30:31.6132740Z ##[debug]contents=**
2022-05-04T22:30:31.6134440Z ##[debug]sourceFolder=/Users/roger/mbp-3/_w/357/a
2022-05-04T22:30:31.6135210Z ##[debug]check path : /Users/roger/mbp-3/_w/357/a
2022-05-04T22:30:31.6136040Z ##[debug]targetFolder=2022-05-04_152740/Generic
2022-05-04T22:30:31.6137970Z ##[debug]cleanTargetFolder=false
2022-05-04T22:30:31.6138510Z ##[debug]overwrite=true
2022-05-04T22:30:31.6139450Z ##[debug]failOnEmptySource=false
2022-05-04T22:30:31.6140450Z ##[debug]flattenFolders=true
2022-05-04T22:30:31.6164920Z Setting up SSH service connection to remote host remoteserver.
2022-05-04T22:30:33.7142640Z ##[debug]include content pattern: **
2022-05-04T22:30:33.7146030Z ##[debug]findPath: '/Users/roger/mbp-3/_w/357/a'
2022-05-04T22:30:33.7147480Z ##[debug]findOptions.allowBrokenSymbolicLinks: 'false'
2022-05-04T22:30:33.7148760Z ##[debug]findOptions.followSpecifiedSymbolicLink: 'true'
2022-05-04T22:30:33.7150040Z ##[debug]findOptions.followSymbolicLinks: 'true'
2022-05-04T22:30:33.7153500Z ##[debug]  /Users/roger/mbp-3/_w/357/a (directory)
2022-05-04T22:30:33.7160760Z ##[debug]  /Users/roger/mbp-3/_w/357/a/MobileBuild.iOS.ipa (file)
2022-05-04T22:30:33.7161630Z ##[debug]2 results
2022-05-04T22:30:33.7162410Z ##[debug]counted 1 files in the source tree
2022-05-04T22:30:33.7164720Z ##[debug]Include matching /Users/roger/mbp-3/_w/357/a/**
2022-05-04T22:30:33.7184760Z ##[debug]Include matched 1 files
2022-05-04T22:30:33.7185600Z ##[debug]Number of files to copy = 1
2022-05-04T22:30:33.7186910Z ##[debug]filesToCopy = /Users/roger/mbp-3/_w/357/a/MobileBuild.iOS.ipa
2022-05-04T22:30:33.7187590Z Found 1 files to copy to the remote machine.
2022-05-04T22:30:33.7188490Z ##[debug]fileToCopy = /Users/roger/mbp-3/_w/357/a/MobileBuild.iOS.ipa
2022-05-04T22:30:33.7189250Z ##[debug]relativePath = MobileBuild.iOS.ipa
2022-05-04T22:30:33.7190540Z Copying file /Users/roger/mbp-3/_w/357/a/MobileBuild.iOS.ipa to ./2022-05-04_152740/Generic/MobileBuild.iOS.ipa on remote machine.
2022-05-04T22:30:33.7192250Z ##[debug]Upload /Users/roger/mbp-3/_w/357/a/MobileBuild.iOS.ipa to ./2022-05-04_152740/Generic/MobileBuild.iOS.ipa on remote machine.
2022-05-04T22:30:34.1142280Z ##[error]Failed to copy /Users/roger/mbp-3/_w/357/a/MobileBuild.iOS.ipa. Unable to create target folder ./2022-05-04_152740/Generic.
**2022-05-04T22:30:34.1154240Z ##[debug]Processed: ##vso[task.issue type=error;]Failed to copy /Users/roger/mbp-3/_w/357/a/MobileBuild.iOS.ipa. Unable to create target folder ./2022-05-04_152740/Generic.**
2022-05-04T22:30:34.1155020Z Completed copying 1 files to the remote machine.
2022-05-04T22:30:34.1155480Z ##[debug]task result: Failed
2022-05-04T22:30:34.1156190Z ##[error]Failed to copy 1 files
2022-05-04T22:30:34.1156760Z ##[debug]Processed: ##vso[task.issue type=error;]Failed to copy 1 files
2022-05-04T22:30:34.1158230Z ##[debug]Processed: ##vso[task.complete result=Failed;]Failed to copy 1 files
2022-05-04T22:30:34.1159450Z ##[debug]Closing the client connection
2022-05-04T22:30:34.1160030Z ##[debug]Task successfully accomplished
2022-05-04T22:30:34.1778090Z ##[section]Finishing: Securely copy files to the remote machine

## Troubleshooting

1. Verified existence and path of file I am trying to copy
2. Verified permissions of file I am trying to copy
3. Verified the task is successfully logging into remote server using service connection
4. Verified the credentials defined in the service connection has access to target server and read/write access to target directories
5. built a new Mac EC2 on Mac OS 11.6.1 and installed vsts agent, ran same pipeline and hit the same issue.
6. I tried creating a new YAML pipeline and attempted to pin task version to a different minor version such as CopyFilesOverSSH@0.175.0, but the pipeline fails YAML linting so I cannot save it.
7. Issue only happened after upgrade to tfs on-prem and agent version upgrade.
8. Tried enabling/disabling "Target machine running Windows", issue persists
9. I plan on upgrading our TFS on-prem version to a higher version such as 18.181.32118.5 if I have to

Is there a way to force downgrade my agent version?  Is there a way to force a different minor version of the task?

**Comment by anatolybolshakov** [comment]
Hi @rogfut there was a significant leap in agent/task versions during your upgrade, so it's hard to tell if this is some agent/task issue or side-effect of some updates at the moment. The issue you mentioned was related to UNC paths, but as I see per logs - you don't use UNC paths...
Seems like task is unable to create target folder on remote machine - does the agent user have enough permissions to create new folders on remote machine via ssh session?

**Comment by rogfut** [comment]
Hi @anatolybolshakov thanks for the reply.  Correct, we don't use UNC paths but I referenced the bug because the symptom is the same - not necessarily the root cause.  Yes, the vsts agent has access to ssh to the target server and has rwx on the root directory where it is trying to write the files.  I confirmed that the vsts agent is logging into the remote server via ssh by looking at the access logs on the remote.  I also used the same credentials that the service connection is using and logged in via winscp client and was able to successfully create directories and write files.

I am actually planning to upgrade my Azure Devops Server to the latest 2020 patch this evening in hopes that a higher agent version / task version fixes the issue.

This is my task yaml if it is helpful

```
steps:
- task: CopyFilesOverSSH@0
  displayName: 'Securely copy files to the remote machine'
  inputs:
    sshEndpoint: '$(Parameters.sshEndpoint)'
    sourceFolder: '$(build.artifactstagingdirectory)'
    contents: '**/*'
    targetFolder: '$(UploadDateTime)/$(TargetFolder)'
```

**Comment by anatolybolshakov** [comment]
Btw have you tried other target folder locations? Does it happen only for specific one?
As I see there is a lack of error details for target folder creation - created [PR](https://github.com/microsoft/azure-pipelines-tasks/pull/16324) to fix this.
I suspect that there could be some other error not related to permissions - could you please also try to copy files to some already existing folder on remote machine, so task would omit target folder creation step, and we would be able to see if there will be any errors during files copying?

**Comment by rogfut** [comment]
Hey @anatolybolshakov thank you for the PR.  I just tried to copy to an existing directory `roger/test`  on the target server and was successful. Additionally, I tried using `ssh2-sftp-client@^5.1.4` - the same lib the task uses - and can successfully create a directory on the target server.  Here is my code:

```
let Client = require('ssh2-sftp-client');

const config = {
  host: 'hostname',
  username: 'username',
  port: 22,
  password: '************',
  keepaliveInterval: 10000
};

let targetPath = 'roger1/test'
let sftp = new Client();

sftp.connect(config)
  .then(() => {
      return sftp.mkdir(targetPath, true);
  })
  .then(data => {
    console.log(data);
  })
  .then(() => {
    sftp.end();
  })
  .catch(err => {
    console.error(err.message);
  })
  ```
  
 I did notice that the CopyFilesOverSSH task does prepend a `./` to the path if it is not absolute before attempting to create it, but that also works for me if I change `let targetPath = 'roger1/test'` to `let targetPath = './roger1/test'`.
 
[relevant code starting at this line](https://github.com/microsoft/azure-pipelines-tasks/blob/872607391d8e3242c2e7cbd3580b2eab0da95196/Tasks/CopyFilesOverSSHV0/copyfilesoverssh.ts#L224)
